"# Here is the journey recording the way of someone like me, growing from a little potato into a little bigger potato."							

"Questions,conversations,characters,numbers"
#Answers, captions, data, short comments
code

#---------2017/6/6------------							
"I have difficulty to reproduct the code in monocle manual"							
HSMM<-estimateDispersions(HSMM)							
disp_table <- dispersionTable(HSMM)							
							
"I updata OS, R and R studio, didn't work."							
#"Finally I realize Ctrapnell mentioned this in his github that:							
# ctrapnell commented on Jun 7 2016							
# estimateDispersions only makes sense right now for negbinomial and negbinomial.size expression families."							
							
#https://github.com/cole-trapnell-lab/monocle-release/issues/5							
							
#"So I can changed expressionFamily=negbinomial.size() and add  lowerDetectionLimit=0.5 while build newCellData							
#It works."							
							
#---------2017/6/9------------							
							
#ggplot tricks:							
text = element_text(size=40)     #larger text including legend title							
legend.text = element_text(size=30) #larger legend text							
legend.key.size = unit(2, "cm") #legend height and width
legend.key.width = unit(2, "cm") #legend width
legend.key.height = unit(2, "cm") #legend height
legend.position="none" #remove legend
							
unit(2, "cm")==grid::unit(1, "in")
							
legend.justification = 'right'						
legend.position=c(0.3,0.82) # change legend position, c() can be negative!"							
guides(colour = guide_legend(override.aes = list(size=10))) #larger legend label							
							
#To control the order and color, change the factor levels to (Epi,Endo, Immune)"
pData(HSMM_3)$CellType <-factor(pData(HSMM_3)$CellType,levels=c("KRT19+ epithelial Cells",
"Stromal+\nEndothelial+\nFibroblasts Cells",
"Immune Cells"))

#CellType and Cluster can be generated by runing reduceDimension()"

#Monocle sequence:
#1)reduceDimension
#2)clusterCells							
#3)plot_cell_clusters							
							
#1)reduceDimension							
#2)orderCells							
#3)plot_cell_trajectory							
							
#---------2017/6/10------------							
#host a Shiny Server/RStudio Server yourself							
#http://deanattali.com/2015/05/09/setup-rstudio-shiny-server-digital-ocean/							
							
#Step 3: Log in to your very own shiny new server							
ssh -l root 67.205.148.56  not  ssh 67.205.148.56 that mentioned in tutoiral							that mentioned in 
su - yah2014							
							
#6.1: Important note re: installing R packages							
sudo su - -c ""R -e \""install.packages('reshape', repos='http://cran.rstudio.com/')\							
sudo su - -c ""R -e \""install.packages('pheatmap', repos='http://cran.rstudio.com/')\						
sudo su - -c ""R -e \""biocLite('monocle', repos='https://bioconductor.org/biocLite.R/')\
'didnt work'

#"log in root and install inside R"							
							
#Shiny Server 32-bit Installation from Source							
#https://www.r-bloggers.com/installing-and-running-shiny-server-from-source-on-32-bit-ubuntu/


#---------2017/6/12 Completed Monocle manual------------


#Monocle manual mistake
'5.2 Finding genes that distinguish cell type or state'
#gene UBC should be the TBP

#---------2017/6/13 Build Shiny------------
rstudio.com/products/shiny/shiny-user-showcase/
bit.ly/shiny-quickstart-2
actionButton(inputId = "go", label = "Update")

#How to save and load S4 dataset
saveRDS(HSMM_epi_3,file = "HSMM_epi_3", ascii = T)
readRDS("HSMM_epi")

#How to generate plot in app
#when as.character turn character into number, use unlist
as.character(unlist(HSMM_ordering_genes))

#How to storage data in dropbox remotely
library(rdrop2)
token <- readRDS("droptoken.rds")
drop_acc(dtoken = token)
outputDir<-"/public/olivier/r/single_cell/rshiny/shinyapps/"
HSMM_ordering_genes <-drop_read_csv(paste0(outputDir,"HSMM_ordering_genes.csv"), dtoken=token)
HSMM_ordering_genes<-as.character(unlist(HSMM_ordering_genes))

drop_read_RDS <- function(file, dest = tempdir(), dtoken = get_dropbox_token(), ...) {
localfile = paste0(dest, "/", basename(file))
drop_get(file, local_file = localfile, overwrite = TRUE, dtoken = dtoken)
readRDS(localfile, ...) #copy from drop_read_csv
}
HSMM_epi_3<-drop_read_RDS(paste0(outputDir,"HSMM_epi_3"), dtoken=token)

#How to setup upload UI

#---------2017/6/14 Extract data from S4 and rebuild DESeq------------
"simpler than I thought"
"while running estimateSizeFactors, the error keeps coming:
Error in estimateSizeFactorsForMatrix(counts(object), locfunc = locfunc,  :
every gene contains at least one zero, cannot compute log geometric means
"
#run this script:
cts <- counts(ddsHTSeq)
geoMeans <- apply(cts, 1, function(row) exp(sum(log(row[row != 0]))/length(row)))
dds <- estimateSizeFactors(ddsHTSeq, geoMeans=geoMeans)

#---------2017/6/15 multiple level of DESeq design------------
#https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#contrasts


res_cluster_2vs1 <- results(dds,contrast = c("Cluster","2","1"), alpha=0.5, lfcThreshold=1)
res_cluster_3vs1 <- results(dds,contrast = c("Cluster","3","1"), alpha=0.5, lfcThreshold=1)
res_cluster_3vs2 <- results(dds,contrast = c("Cluster","3","2"), alpha=0.5, lfcThreshold=1)

#Some useful function for S4 vector
isS4()
class?ExpressionSet
HSMM_epi_3@phenoData =phenoData(HSMM_epi_3)
new_object <- updateObject(old_object)
showMethods("as.data.frame")
getMethod("as.data.frame", "DataFrame")
method?as.data.frame
DataFrame


#---------2017/6/16 dds vs dds_results------------
#generate plotCounts plot 1:9 ggplot
d<-data.frame(x= str(0), y= integer(0))
for( i in 1:9){
d<-plot(,returnData=T)
d<- rbind(d,d1)
}
ggplot(d)+facet_warp(~gene_short_name)

#facet_warp vs facet_grid

#generate Rshiny
"print(p), but forgot p<-ggplot"

#"Before publish to shiny, restart the R to double check the package dependency！！！"

#//
#//  Debug.R
#//  
#//
#//  Created by Yang Hu on 2017/6/20.
#//
#//

#include <stdio.h>

sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)
#Error in datk[c(startG:endG), ] = foreach(t = actualThreads, .combine = rbind) %dopar%  :
#number of items to replace is not a multiple of replacement length
#"The code below uses parallel computation where multiple cores are available. This works well when R is run from a terminal or from the Graphical User Interface (GUI) shipped with R itself, but at present it does not work with RStudio and possibly other third-party R environments." -WGCNA package tutorial




#Import data into R with an unknown number of columns?
#You need to use the fill argument in the read.table function
fill=TRUE
#There is nice function count.fields (see help) which counts number of column per row:
no_col <- max(count.fields("test", sep = "\t"))
data <- read.table("test",sep="\t",fill=TRUE,col.names=1:no_col)
data

#remove header which have "!"
#grepl returns a logical vector
grepl

#
grep("!Sample_title",MicroArrayData0[,1])
MicroArrayData0[,1]=="!Sample_title"



#the factor variable retains all of its original levels -- even when they do not exist in the new data frame.
droplevels

#Error in sort.list(y) : 'x' must be atomic for 'sort.list'
#Have you called 'sort' on a list?
unlist

#split column into multiple ones
str_split_fixed

#remove rows with empty value in column 1
X <- X[!X[,1]=="",]

#Increase momery
memory.limit(size=65000)

#---2017-07-01
"What is the difference between gsub and sub methods in r"
#The g stands for global, as in replace globally (all):

sub('l', '*', "hello")  #=> "he*lo"
gsub('l', '*',"hello")  #=> "he**o"

#sub(".*") can remove undefined content
filename <- c("GSM726928_OOC1.CEL.gz","GSM726949_STC2_1.CEL.gz")
sampleNames <- sub(".*_", "", filename) #=> "OOC1.CEL.gz" "1.CEL.gz"
sampleNames <- sub(".CEL.gz$", "", sampleNames) #=> "OOC1" "1"


#Avoid rbind()/cbind() conversion from numeric to factor
cbind.data.frame

#merge unequal dataframes and replace missing rows with 0
merge(df1, df2, all = TRUE)

#adjust all log data to positive
logcounts0 <- cpm(expr0,log=TRUE)
min(logcounts0)
all((logcounts0+5.2)>0)

#What does %>% function mean in R?
#The dplyr package introduced the %.% operator to pass the left hand side
#as an argument of the function on the right hand side, similar to a *NIX pipe. |
df <- data.frame(
        x = sample(10, 10, rep = TRUE),
        y = sample(10, 10, rep = TRUE),
        z = sample(10, 10, rep = TRUE)
)
nrow(df)
nrow(distinct(df))
nrow(distinct(df, x, y))
nrow(distinct(df, x))

distinct(df, x)
distinct(df, y)

# Can choose to keep all other variables as well
distinct(df, x, .keep_all = TRUE)
distinct(df, y, .keep_all = TRUE)

# You can also use distinct on computed variables
distinct(df, diff = abs(x - y))

# The same behaviour applies for grouped data frames
# except that the grouping variables are always included
df <- tibble(
        g = c(1, 1, 2, 2),
        x = c(1, 1, 2, 1)
) %>% group_by(g)
df %>% distinct()
df %>% distinct(x)


#To remove the subtitle use the following:
plot(hc, xlab="", sub="")

#----------Task on 2017/07/05----------------
#try to remove genes that don't' change much in the RNAseq experiment
thresh <- myCPM > 0.5 #thresh number decide lower limits of std
keep <- rowSums(thresh) >= 9 # keep rowsum number decide individule circles

#make a heatmap out of the correlation matrix c:
#which microarray sample is each of Danwei's experiments closest to
pheatmap(cor(rnaseq_MicroArrayData, method="spearman"))
#pickup gene set with p value <0.05
#Do deseq, 

#find out linkID
#complete WGCNA


#Plot a data frame as a table in r
library(gridExtra)
title <-as.data.frame(eData$title)
dev.off()
grid.table(title)

# =====================================================================================
# 
#  4. DEseq  ----some values in assay are not integers
# 
# =====================================================================================
# read count matrix
rnaseq_MicroArrayData <-read.csv("rnaseq_MicroArrayData.csv", row.names = 1)

# create sample table for DESeq
sample.table <- data.frame( title = colnames(rnaseq_MicroArrayData),
                            group = c(rep("Danwei",9),
                                      as.character(eData$labelDescription))
)
dim(sample.table) 
## Creating a DESeqDataSet object
#create ddsHTSeq
ddsHTSeq <- DESeqDataSetFromMatrix( countData = rnaseq_MicroArrayData,
                                    colData = sample.table,
                                    design= ~ group)
ddsHTSeq


#Combining data frames with unequal number of columns [duplicate]
df1 = data.frame(A = c(ncol(rnaseq_MicroArrayData),2,1))
df2 = data.frame(B = c("#","Danwei","StemCell","StemCellLine"))
df3 = data.frame(C=c(rep("Danwei",9),as.character(eData$labelDescription)))
l <- merge(t(df1),t(df2),all=TRUE)
l <- merge(l,t(df3),all=TRUE)


#----------Task on 2017/07/07
#try to keep the genes that have higher std (top 25%) in affrymetric data
sd = apply(rnaseq_MicroArrayData[,10:ncol(rnaseq_MicroArrayData)], 1, sd)
sd = sd[order(sd, decreasing=TRUE)]
Top_Std_genes <-names(sd)[1:(length(sd)*0.25)]

# The las argument rotates the axis names
par(las=2)

#similarity of signatures
#find out linkID
#complete WGCNA

#----------Task on 2017/07/08
#add trend line in plot r
lines(lowess(median,sd), col="blue")


#generate table in R markdown
cellType.table <- read.delim("CellType.txt")
kable(cellType.table, captioin ="Cell Type")


#generate table in R
x <- data.frame(row.names=paste("Name",1:10))
x[,1] <- 1:10
x[,2] <- sample(1:100,10)
x[,3] <- sample(LETTERS[1:26],10)
colnames(x) <- c("Value 1", "Value 2", "Label")

#Plot your table with table Grob in the library(gridExtra)
ss <- tableGrob(x)

#Make a scatterplot of your data
k <- ggplot(x,aes(x=x$"Value 1",y=x$"Value 2")) + 
        geom_point()

#Arrange them as you want with grid.arrange
grid.arrange(ss)

#2017-07-14
#try NK cells and other myloide cells

#CD3G(-),NCAM1(+),KLRD1 (+),NCR1(+)	, CD3-, CD56+, CD94+, NKp46+

#assaign each cluster to cell type

#2017-07-18
#FlowJo pineer in single cell level clustering using machine learning
#k-means clustering, tsne 
#http://www.the-scientist.com/?articles.view/articleNo/49228/title/Picking-Out-Patterns/
#https://followthedata.wordpress.com/2015/12/21/list-of-deep-learning-implementations-in-biology/
ADAGE #Analysis using Denoising Autoencoders for Gene Expression 
#https://github.com/greenelab/adage

#2017-0720
#counts non-zero data
apply(exprs(gbm_cds), 2, function(c)sum(c!=0)) 

#ggplot range
g+coord_cartesian(xlim = c(0, 30000))
g+coord_cartesian(xlim = c(0, 10000))

#density ggplot
qplot(counts_per_cell, data=pData(gbm), geom="density")

#2017-07-21
#lazy-load database 'C:/Users/User/Documents/R/win-library/3.4/VGAM/R/VGAM.rdb' is corrupt
remove.packages("VGAM")
install.packages("VGAM")

#`maximal number of DLLs reached.
#stop loading large amount of libraries

#ggplot
#http://ggplot2.tidyverse.org/reference/index.html#section-plot-basics

#2017-07-24
#gbm for counts
#sce for fpkm

#Error: Aesthetics must be either length 1 or the same as the data (1397): x, y

#2017-07-30-----------------
"how to store variable to list in for loop?"
#list all files with a specified extension
tab.files <-list.files(pattern = "\\.tab$") 
#creat a empty list
tabfileslist <- list() 
for(i in 1:n){
        tab <- tab.files[i] #get file name
        #store variable to list
        tabfileslist[[tab]] <- read.delim(tab) 
        #select 100 genes and 4 columns
        tabfileslist[[tab]] <- tabfileslist[[tab]][1:100,select_columns]
        #filter out genes with RNA.TS score not greater than 0
        tabfileslist[[tab]] <-subset(tabfileslist[[tab]],tabfileslist[[tab]][,"RNA.TS"]>0)
        print(dim(tabfileslist[[tab]]))
}

"How to trim multiple strings"
celltype <- gsub('.tab', '',tab.files)  #"B_Cells.tab" => "B_Cells"

"How to make heatmap with groupedrows and columns"
#https://stackoverflow.com/questions/29740957/how-to-create-pre-annotated-rowside-column-in-heatmap-2
#https://stackoverflow.com/questions/41648293/r-pheatmap-perform-clustering-and-show-dendrograms-per-annotation-category

        "1). pheatmap doesn't work"
#skip all Arguments and add back one by one
#delete arguments filename = "TEST_12cat.png"....

        "2).R list to data frame without lossing any information"
do.call(rbind.data.frame, tabfileslist)
        
        "3). Make a column with duplicated values unique in a dataframe"
make.unique(as.character(myexprs$Gene), sep = ".")
        